### PR TITLE
Pluggable lingual parser

### PR DESCRIPTION
**Pull request by HiromuHota**
This PR adds 
-  `LingualParser` class, which any lingual parser like `Spacy` should inherit from
- `lingual_parser` as an argument to `Parser` to be able to plug a custom lingual parser
Unless otherwise specified, `Spacy` will be used by default for backward-compatibility.

### CODE REVIEW

File: src/fonduer/parser/lingual_parser.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+        :param text: text to be split
+        :type text: str
+        :return: A generator of dict that is used as `**kwargs` to instantiate
+            :class:`Sentence <fonduer.parser.models.Sentence>`.
+        """
+        raise NotImplementedError()
+
+    def enrich_sentences_with_NLP(
+        self, all_sentences: Iterable[Sentence]
```
Comment (SenWu) [inline src/fonduer/parser/lingual_parser.py]: Just use `sentences` ?

File: src/fonduer/parser/spacy_parser.py [hunk 2]
```diff
@@ -15,7 +17,7 @@
     raise Exception("spaCy not installed. Use `pip install spacy`.")
 
 
-class Spacy(object):
+class Spacy(LingualParser):
```
Comment (SenWu) [inline src/fonduer/parser/spacy_parser.py]: Rename to `SpacyParser`?

File: src/fonduer/parser/lingual_parser.py (hunk 1 above)
Comment (HiromuHota) [reply src/fonduer/parser/lingual_parser.py]: Renamed into `sentences` by 0d480b893a3758447579c40c3ae6b252957a31dc.

File: src/fonduer/parser/spacy_parser.py (hunk 2 above)
Comment (HiromuHota) [reply src/fonduer/parser/spacy_parser.py]: Renamed as suggested at b07538a706aa03a33ebda4510d2144309da7dc28.

**Comment by HiromuHota** [comment]
Thanks for the suggestions. I'm currently adopting all of them.
Will push them when I'm done.

**Comment by HiromuHota** [comment]
Rebased and (force) pushed. Let me know if there is anything.

File: CHANGELOG.rst [hunk 3]
```diff
@@ -57,6 +57,9 @@ Added
         from fonduer.parser import Parser
         parser = Parser(session, vizlink=CustomVisualLinker())
 
+* `@HiromuHota`_: Add `LingualParser`, which any lingual parser like `Spacy` should inherit from,
+  and add `lingual_parser` as an argument to `Parser` to be able to plug a custom lingual parser.
+
```
Comment (SenWu) [inline CHANGELOG.rst]: Can we update CHANGELOG as well?

Comment (HiromuHota) [reply CHANGELOG.rst]: I added some changes to CHANGELOG but I didn't include changes that do not affect how to use Fonduer like additional tests, doc section, etc.

**Comment by HiromuHota** [comment]
I believe the following question is for #290. Let me know if I'm wrong.

> One more question about typing: do you think it's a good idea to have type folder to include all types in Fonduer?

**Comment by SenWu** [comment]
Yes, you are right!

**Pull request merged**